File: python/src/coffeeshop_kata/order.py
from .status import Status


class Order:
    def __init__(self, order_id: int) -> None:
        self.id: int = order_id
        self.status: Status | None = None

    def set_status(self, status: Status) -> None:
        self.status = status

    def is_update_allowed(self) -> bool:
        return not self.is_paid()

    def is_cancel_allowed(self) -> bool:
        return not self.is_paid()

    def is_paid(self) -> bool:
        """
        BUG! All statuses except PAYMENT_EXPECTED mean that it is paid.
        Off by one error: Status.PAID should also return True.
        """
        if self.status is None:
            return False
        # This comparison preserves the exact same bug as Java version
        # Uses > instead of >= causing PAID status to incorrectly return False
        status_order = list(Status)
        return status_order.index(self.status) > status_order.index(Status.PAID)

File: typescript/src/order.ts
import { Status } from "./status";

export class Order {
    public readonly id: number;
    public status?: Status;

    constructor(id: number) {
        this.id = id;
    }

    setStatus(status: Status): void {
        this.status = status;
    }

    isUpdateAllowed(): boolean {
        return !this.isPaid();
    }

    isCancelAllowed(): boolean {
        return !this.isPaid();
    }

    // BUG! All statuses except PAYMENT_EXPECTED mean that it is paid.
    // Off by one error: PAID should also return true.
    isPaid(): boolean {
        if (!this.status) {
            return false;
        }
        const statusOrder = Object.values(Status);
        const currentIndex = statusOrder.indexOf(this.status);
        const paidIndex = statusOrder.indexOf(Status.PAID);
        return currentIndex > paidIndex;
    }
}
